Basic DNS seed server implementation

pavel-kokolemin opened the pull request:
Add a basic scanner/DNS server that nodes can use for initial seeding.

To keep things simple, the server will try to keep connections open to all reachable nodes.
When a new outbound connection is made, a new DNS record is added (but only for nodes on default ports).
When the connection is closed, the DNS record is removed.
When a connection fails, the server increases the backoff time between connection attempts.
If the number of failed connection attempts exceeds the limit, the address is removed from the list.
Once-reachable and newer-reachable addresses have different connection failure limits
(equivalent to about 1 month and about 1 hour, respectively).

I'm attaching sample DNS responses from this server and from Bitcoin's ([https://github.com/sipa/bitcoin-seeder](https://github.com/sipa/bitcoin-seeder)):
[mintlayer.txt](https://github.com/mintlayer/mintlayer-core/files/10674942/mintlayer.txt)
[bitcoin.txt](https://github.com/mintlayer/mintlayer-core/files/10674943/bitcoin.txt)

TheQuantumPhysicist commented on p2p/src/config.rs:
```diff
@@ -19,6 +19,7 @@ use utils::make_config_setting;
 use crate::net::types::PubSubTopic;
 
+// TODO: Use different ports on testnet and regtest
 pub const DEFAULT_BIND_PORT: u16 = 3031;
```
The ports should be taken from ChainConfig

pavel-kokolemin replied:
OK, thanks. I added `p2p_port` field to `ChainConfig`.

TheQuantumPhysicist replied:
I really remember it was there before... I wonder where it went. Anyway, thank you.

TheQuantumPhysicist commented on dns_server/src/server/mod.rs:
```diff
@@ -0,0 +1,286 @@
+    command_tx: mpsc::UnboundedReceiver<ServerCommands>,
+}
+
+// Same values as in `https://github.com/sipa/bitcoin-seeder/blob/master/dns.cpp`
```
Always use permalinks in such situations

pavel-kokolemin replied:
Sure, updated.

TheQuantumPhysicist commented on dns_server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+    pub mbox: String,
+}
+
+const DEFAULT_DATA_DIR_NAME: &str = ".mintlayer";
```
Is it normal to use the same data directory for the node and the dns server?

pavel-kokolemin replied:
I switched to using `BaseDirs::data_dir` instead (on Linux this is `$HOME/.local/share/mintlayer_dns_server`).

TheQuantumPhysicist replied:
I would rather keep it in $HOME... I hate it when things go into .local/share. That's a personal preference though... is there a reason for going in there?

pavel-kokolemin replied:
There is a specification for this: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
Some random article from the Internet: https://0x46.net/thoughts/2019/02/01/dotfile-madness/
And it's mentioned in our issues: https://github.com/mintlayer/mintlayer-core/issues/498

Maybe we should just remove default from the `datadir` argument (make it mandatory)?

TheQuantumPhysicist replied:
I never realized people hated this... I actually like it. But sure, let's follow that standard.

But no, we can't just force people to use that directory without the option to put it somewhere else. There can be all kinds of issues, like diskspace or permission issues.

TheQuantumPhysicist commented on dns_server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+    #[clap(long)]
+    pub ns: String,
+
+    /// E-Mail address reported in SOA records (RNAME)
```
Well, I'd be too hesitant to enforce using a valid email address... I'm always worried about spam.

pavel-kokolemin replied:
Looks like the SOA record must always be present: https://serverfault.com/a/1024024
And `rname` is the admin's email address in the SOA.
It's possible to run DNS without the SOA record, but all of Bitcoin's public seed DNSs have it, so I decided to make it mandatory.

TheQuantumPhysicist replied:
I would want to keep it optional and use a bogus email address if none is provided. This isn't the 90s anymore.

Can we technically make it optional?

pavel-kokolemin replied:
Sure, I made both the `nameserver` and `mbox` arguments optional (the same way it's done in bitcoin-seeder).

TheQuantumPhysicist commented on dns_server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+    pub datadir: String,
+
+    /// IP address and UDP port to listen on
+    #[clap(long, default_values_t = vec!["[::]:53".to_string()])]
```
Why not 0.0.0.0:53?

pavel-kokolemin replied:
On Linux, `[::]` works with both IPv6 and IPv4 clients (by default).

TheQuantumPhysicist commented on dns_server/src/main.rs:
```diff
@@ -0,0 +1,84 @@
+    let base_transport = TcpTransportSocket::new();
+    let transport = NoiseTcpTransport::new(stream_adapter, base_transport);
+    let chain_config = Arc::new(common::chain::config::create_mainnet());
+    let p2p_config = Default::default();
```
Do we really have to instantiate a full Mintlayer node server? Can't we get away with simple pinging that port? Or just opening a connection and closing it?

pavel-kokolemin replied:
I'm using the generic `NetworkingService` from p2p, which works like a low-level networking implementation.
It's definitely not a full node server.
Simply pinging the ports is not enough, because we also need to know when new nodes join the network.

TheQuantumPhysicist commented on dns_server/src/main.rs:
```diff
@@ -0,0 +1,84 @@
+async fn run(config: Arc<DnsServerConfig>) -> Result<void::Void, error::DnsServerError> {
+    let (command_tx, command_rx) = mpsc::unbounded_channel();
+
+    let stream_adapter = NoiseEncryptionAdapter::gen_new();
```
Please do all the instantiations of p2p in a separate function and take the result and use it. There seems to be many steps that shouldn't be done in here.

pavel-kokolemin replied:
OK, updated.

TheQuantumPhysicist commented on dns_server/src/server/mod.rs:
```diff
@@ -0,0 +1,286 @@
+    DelAddress(IpAddr),
+}
+
+pub struct Server {
```
Why do we call this Server, and not DnsServer? I would prefer the latter unless there's a good reason not to do it.

pavel-kokolemin replied:
Sure, updated.

TheQuantumPhysicist commented on dns_server/src/server/mod.rs:
```diff
@@ -0,0 +1,286 @@
+fn handle_command(auth: &AuthorityImpl, command: ServerCommands) {
+    match command {
+        ServerCommands::AddAddress(IpAddr::V4(ip)) => {
+            auth.ip4.lock().expect("mutex must be valid").push(ip);
```
These expects are too aggressive... deeming the whole program dead just because a wrong format was submitted. Perhaps we should just return an error with one shot.

TheQuantumPhysicist commented on dns_server/src/server/mod.rs:
```diff
@@ -0,0 +1,286 @@
+        let ipv4 = self
+            .ip4
+            .lock()
+            .expect("mutex must be valid")
```
Perhaps a dumb comment... but please change "mutex must be valid" to "mutex must be valid (ipv4)". Using the same message in multiple places can lead to confusion when an error happens.

pavel-kokolemin replied:
Sure, updated.

TheQuantumPhysicist commented on dns_server/src/server/mod.rs:
```diff
@@ -0,0 +1,286 @@
+    ns: Name,
+    mbox: Name,
+    inner: InMemoryAuthority,
+    ip4: Mutex<Vec<Ipv4Addr>>,
```
Please do not use std::Mutex in async code. Using tokio mutex and away on it.

pavel-kokolemin replied:
Std `Mutex` is OK here because the lock times are short.
From https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html:
> Contrary to popular belief, it is ok and often preferred to use the ordinary [Mutex](https://doc.rust-lang.org/nightly/std/sync/mutex/struct.Mutex.html) from the standard library in asynchronous code.

TheQuantumPhysicist replied:
It's a bad habit and can lead to deadlocks. It's hard here because it's just guarding data, but as per our rules of thumb we don't this because people misunderstand it. But fine. Let's go with it here.

TheQuantumPhysicist commented on dns_server/src/crawler/storage_impl.rs:
```diff
@@ -0,0 +1,94 @@
+    /// Database schema for peer db storage
+    pub Schema {
+        /// Table for all reachable addresses
+        pub DBAddresses: Map<String, ()>,
```
Please add a schema for versioning too. You need to store the version of your schema.

pavel-kokolemin replied:
Sure, added.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,516 @@
+        true
+    }
+
+    /// Address list maintaince
```
maintaince?

pavel-kokolemin replied:
Fixed, thanks.

TheQuantumPhysicist commented:
Very nice work.

This is great, but we're still missing many things. Besides the comments I made, I think there are tons of tests that are still needed before getting this to be merged.

pavel-kokolemin commented on dns_server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+
+const DEFAULT_DATA_DIR_NAME: &str = ".mintlayer";
+
+fn default_data_dir() -> String {
```
I could not use `PathBuf` with `clap::Parser` because `PathBuf` does not implement `std::fmt::Display`.

TheQuantumPhysicist replied:
That's unfortunate.

pavel-kokolemin commented on dns_server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+
+    /// IP address and UDP port to listen on
+    #[clap(long, default_values_t = vec!["[::]:53".to_string()])]
+    pub bind_addr: Vec<String>,
```
Right, updated, thanks.

pavel-kokolemin commented on dns_server/src/config.rs:
```diff
@@ -0,0 +1,58 @@
+
+    /// Hostname of the nameserver
+    #[clap(long)]
+    pub ns: String,
```
Updated, thanks.

pavel-kokolemin commented:
> Very nice work.
> 
> This is great, but we're still missing many things. Besides the comments I made, I think there are tons of tests that are still needed before getting this to be merged.

Thanks! I'm working on tests.

TheQuantumPhysicist commented on dns_server/src/config.rs:
```diff
@@ -0,0 +1,61 @@
+
+    /// Use testnet
+    #[clap(long)]
+    pub testnet: bool,
```
Please don't use boolean for this. We may have to instantiate multiple testnets. The right thing here is to use network ids that parse from strings. So "testnet" is testnet, and "mainnet" is mainnet. The default is mainnet.

pavel-kokolemin replied:
OK, updated. `clap::ValueEnum` worked nicely.

TheQuantumPhysicist commented on dns_server/src/main.rs:
```diff
@@ -0,0 +1,94 @@
+
+    let chain_config = if config.testnet {
+        Arc::new(common::chain::config::create_testnet())
+    } else {
```
And after using network ids, we match each possible string. If none matches, we panic.

pavel-kokolemin replied:
Should be OK now.

TheQuantumPhysicist commented on dns_server/src/main.rs:
```diff
@@ -0,0 +1,94 @@
+mod dns_server;
+mod error;
+
+fn make_p2p_transport() -> NoiseTcpTransport {
```
I think this can carry risk. The decision of transport should be made in p2p and exported here. This is code duplication.

The bad thing with this is that in the future, if we change the wrapping protocol, this will break the DNS server.

pavel-kokolemin replied:
Updated, thanks.

pavel-kokolemin commented:
I added unit tests.

There is a risk that at some point the DNS server will not conform to the network protocol.
For example, by not responding to some requests from the sync manager (currently all requests are ignored).
After that, it would be banned by all nodes and stop working properly. I don't know what to do about it.

pavel-kokolemin commented:
I forgot to mention that with these changes the node will not bind to the listening port (unless explicitly specified with `--p2p-addr`). Curiously enough, all tests still pass.
I guess I should revert this first. I will mark the PR as draft for now.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+        let tx = storage.transaction_ro()?;
+
+        let storage_version = tx.get_version()?;
+        if storage_version.is_some() && storage_version != Some(STORAGE_VERSION) {
```
Why not use if let or match? This is more like the C++ way of doing things...

pavel-kokolemin replied:
Sure, updated.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+        storage: S,
+        command_tx: mpsc::UnboundedSender<ServerCommands>,
+    ) -> Result<Self, DnsServerError> {
+        // Load all persistent addresses
+        let tx = storage.transaction_ro()?;
+
+        let storage_version = tx.get_version()?;
+        if storage_version.is_some() && storage_version != Some(STORAGE_VERSION) {
+            return Err(DnsServerError::Other("Unexpected storage version"));
+        }
+
+        let mut addresses = BTreeMap::new();
+        for address in tx.get_addresses()?.iter().filter_map(|address| address.parse().ok()) {
+            Self::new_address(&mut addresses, address, true);
+        }
+        tx.close();
+
```
Please move all the selected code into a private function and return `addresses` from that.

TheQuantumPhysicist replied:
This code can be reordered to separate concerns much better than it looks now. I see storage_version is used later... so my exact suggestion from above isn't directly possible, but this can still be improved. Just know that whenever I see `mut` and then usage the alarms go off in my head.

pavel-kokolemin replied:
Right. Updated. I hope it's OK now.

pavel-kokolemin replied:
Updated one more time.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+    ) {
+        match request {
+            PeerManagerRequest::AddrListRequest(_) => {
+                // Ignored
```
Why not provide an address list? I don't know if it's the right thing to do, but I'm interested in looking into this as a possibility.

pavel-kokolemin replied:
For security the nodes only load addresses (in a list) from outbound connections and the DNS server does not accept inbound connections at all. So normally nobody would do that. And the DNS server has other means to sharing addresses (via the DNS protocol).
Later we might implement loading addresses from the nodes to the server.

TheQuantumPhysicist replied:
What are the security concerns?

pavel-kokolemin replied:
This is because inbound peers should be treated as more dangerous than outbound peers. An attacker could easily target a specific node if it accepts inbound connections, connecting to it from different IPs, for example.

TheQuantumPhysicist replied:
Gotcha.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+                // Ignored
+            }
+            PeerManagerRequest::AnnounceAddrRequest(AnnounceAddrRequest { address }) => {
+                // TODO: Rate limit AnnounceAddrRequest requests
```
Please add more information, as to why rate limit (what are the concerns) and against what (ip addresses?)

pavel-kokolemin replied:
Sure, done.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+            .addresses
+            .get_mut(&address)
+            .expect("address in the Connecting state must be known");
+        let valid = peer_info.network == self.config.network;
```
This doesn't seem to be a good strategy to do this check. The PeerInfo object should have something like `peer_info.is_compatible(&other_peer_info)` (or whatever else you find appropriate). There should not be any processing that's independent of p2p. It doesn't make sense that we're using the mintlayer p2p protocol, yet we invent our own rules here. Consider the effect of updates. 

I see here we're dealing with the crawler config. But being invasive like this is not the way to go. Please study this in a broader scope and let's see if we can come with a better compatibility check than just checking an arbitrary member variable in a struct.

pavel-kokolemin replied:
I added `is_compatible` to `PeerInfo`. For now it only checks `network`, but I added TODO to check version as well.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+        let valid = peer_info.network == self.config.network;
+
+        if !valid {
+            log::debug!("invalid peer detected at {}", address.to_string());
```
this is `info!`, not `debug!`.

pavel-kokolemin replied:
Sure, updated.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+
+        if !valid {
+            log::debug!("invalid peer detected at {}", address.to_string());
+            self.conn.disconnect(peer_info.peer_id).expect("disconnect must succeed");
+            Self::change_address_state(
+                &self.config,
+                &address,
+                address_data,
+                AddressState::Disconnected,
+                &mut self.storage,
+                &self.command_tx,
+            );
+            return;
```
Please put both these calls in a separate function. We don't update states within code logic.

pavel-kokolemin replied:
Sorry, but I don't like this. It seems like extra boilerplate to me. Let's keep it, please.

pavel-kokolemin replied:
I looked one more time and still think that we should keep the code as it is.
That's because connected peers could only ever be created in the `handle_outbound_accepted` function.
So this new proposed function could only be called from one place - `handle_outbound_accepted`.

TheQuantumPhysicist replied:
What if someone in the future needed another way to create/delete peers? Instead of calling this function that does two things, they have to copy two things, hopefully.

pavel-kokolemin replied:
I added `create_outbound_peer` and `remove_outbound_peer` functions. Should be OK now.
Also renamed `peers` to `outbound_peers` to make it clearer that only outbound peers are stored there.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+            return;
+        }
+
+        Self::change_address_state(
+            &self.config,
+            &address,
+            address_data,
+            AddressState::Connected,
+            &mut self.storage,
+            &self.command_tx,
+        );
+        self.peers.insert(peer_info.peer_id, address);
+    }
```
Same here.

pavel-kokolemin replied:
Also same.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+    }
+
+    fn handle_misbehaved(&mut self, _peer_id: N::PeerId, _error: P2pError) {
+        // Ignore all misbehave reports
```
Are we sure? Why not disconnect them or remove them from lists?

pavel-kokolemin replied:
`NetworkingService` normally does not send `Misbehaved` messages (except for `ANNOUNCEMENT_MAX_SIZE` check), so I decided to keep things simpler.

TheQuantumPhysicist replied:
Don't you think this could open the door to flooding attacks or crash the node with huge messages?

pavel-kokolemin replied:
Huge messages are not a problem because the backend limits the maximum message size to 10MB.
Flooding attacks shouldn't be a problem either, because we don't do any processing.
Anyway, I added a TODO and created an issue (https://github.com/mintlayer/mintlayer-core/issues/699) for this.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,560 @@
+        true
+    }
+
+    /// Address list maintenance
```
Please type more description to what's supposed to happen in this function.

pavel-kokolemin replied:
Sure, updated.

TheQuantumPhysicist commented on dns_server/src/dns_server/mod.rs:
```diff
@@ -0,0 +1,287 @@
+            auth.ip6.lock().expect("mutex must be valid (ipv6)").push(ip);
+        }
+        ServerCommands::DelAddress(IpAddr::V4(ip)) => {
+            auth.ip4.lock().expect("mutex must be valid (ipv4)").retain(|val| *val != ip);
```
Again, I see "mutex must be valid (ipv4)" repeated twice. Please make the repeated expect messages distinguishable.

pavel-kokolemin replied:
Updated. I did not pay much attention because `expect` would also print the filename and line number, which is usually enough.

TheQuantumPhysicist replied:
That assumes you know what version you're testing against. It often happens that we don't know what version caused an issue with a user.

TheQuantumPhysicist commented on p2p/src/lib.rs:
```diff
@@ -164,12 +167,29 @@ pub async fn make_p2p<S: PeerDbStorage + 'static>(
+    let transport = make_p2p_transport();
 
-    let p2p = P2p::<DefaultNetworkingService<NoiseTcpTransport>>::new(
+    let bind_addresses = if !p2p_config.bind_addresses.is_empty() {
```
Perhaps this logic should be moved to a separate function.

Also I keep thinking whether we want BindAddresses to be a trait with a Default impl coming from the application... I don't know. But seems like it's "its own thing".

pavel-kokolemin replied:
Sure, updated.

TheQuantumPhysicist reviewed (commented):
Looking much better. I have to say that there are many things that are questionable in the crawler. This doesn't mean I have a better answer. But generally speaking, I don't like seeing these big state machines that just keep getting updated all the time. The crawler impl is large (500+ lines) and is not separated in anyway. Perhaps considering separating it into multiple modules that are tested on their own is a good idea. 

Also the tests have many helpers that are thrown within the tests. Maybe using a module for these is a cleaner way.

Please address the comments and I'll take another look.

pavel-kokolemin commented on dns_server/src/crawler/tests.rs:
```diff
@@ -0,0 +1,419 @@
+async fn advance_time(
+    crawler: &mut Crawler<MockNetworkingService, DnsServerStorageImpl<storage::inmemory::InMemory>>,
+    time_getter: &P2pTestTimeGetter,
+    hours: u32,
```
Sure, updated.

pavel-kokolemin commented on dns_server/src/dns_server/mod.rs:
```diff
@@ -0,0 +1,287 @@
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ServerCommands {
+    AddAddress(IpAddr),
+
+    DelAddress(IpAddr),
+}
```
Sorry, but this seems like a micro-optimization to me. Nodes normally become online/offline one by one, and we are probably talking about maximum of 1000 reachable nodes (I think the bitcoin network has less than 20000 reachable nodes). Let's not spend too much time on this yet.

pavel-kokolemin commented:
> The crawler impl is large (500+ lines) and is not separated in anyway. Perhaps considering separating it into multiple modules that are tested on their own is a good idea.

OK, I will make a note. But let's please keep it for now.

> Also the tests have many helpers that are thrown within the tests. Maybe using a module for these is a cleaner way.

Sure, updated.

TheQuantumPhysicist commented on dns_server/src/crawler/tests/mod.rs:
```diff
@@ -0,0 +1,255 @@
+    (crawler, state, command_rx, time_getter)
+}
+
+async fn advance_time(
```
Can you please provide description for this function that explains how it works?

pavel-kokolemin replied:
Sure, done.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,552 @@
+        }
+        tx.close();
+
+        if storage_version.is_none() {
```
Move into match

pavel-kokolemin replied:
Done.

TheQuantumPhysicist commented on p2p/src/net/types/mod.rs:
```diff
@@ -60,6 +60,13 @@ pub struct PeerInfo<P> {
 }
 
+impl<P> PeerInfo<P> {
+    pub fn is_compatible(&self, network: [u8; 4]) -> bool {
```
The `is_compatible` function should take a config, not a network, where multiple things can be tested.

pavel-kokolemin replied:
Sure, updated.

TheQuantumPhysicist commented on p2p/src/lib.rs:
```diff
@@ -156,6 +151,33 @@ impl subsystem::Subsystem for Box<dyn P2pInterface> {}
+    NoiseTcpTransport::new(stream_adapter, base_transport)
+}
+
+fn get_p2p_bind_addresses(bind_addresses: &[String], p2p_port: u16) -> Result<Vec<SocketAddr>> {
```
A better signature here would be

```
fn get_p2p_bind_addresses<S: AsRef<str>>(bind_addresses: &[S], ...)
```

Then add `as_ref()` to addresses elements, and instead of `clone()`, use `to_owned()`.

pavel-kokolemin replied:
Updated, thanks.

TheQuantumPhysicist commented on dns_server/src/crawler/mod.rs:
```diff
@@ -0,0 +1,567 @@
+    /// Update address state.
+    ///
+    /// The only place where the address state can be updated.
+    fn change_address_state(
```
I think this function should be implemented as a finite state machine. Let's talk about it.

TheQuantumPhysicist replied:
So as agreed, this is gonna be the next PR.

TheQuantumPhysicist reviewed (commented):
This is getting much better. Thank you! Please do the little fixes that are left.

About the state thing, let's talk about it and see how we'll go about it.

Again, apologies for holding this long on the queue, but this is a new program, so it kind of deserves extra scruitiny :smile:

pavel-kokolemin commented:
I added `update_db` helper function to retry DB transactions in case of recoverable errors.

I also noticed some duplicated code when processing untrusted data from the network (`PeerAddress` values).
And some of the checks in the DNS server were missing because of this duplication.
So I refactored the `TransportAddress` trait to also do the validation. It should work, but I'm not sure if this is a good approach.
